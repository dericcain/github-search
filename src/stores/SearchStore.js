import { observable, action } from 'mobx';
import { searchRequest, userInfoRequest, followersRequest } from '../utils/github-request';
import debounce from '../utils/debounce';

class SearchStore {
  @observable searchTerm = '';
  @observable isLoading = false;
  @observable resultSet = [];
  @observable user = null;
  @observable followers = [];
  @observable nextPageOfFollowersLink = null;
  @observable isLoadingMore = false;

  @action
  updateSearchTerm(searchTerm) {
    this.searchTerm = searchTerm;
  }

  // We want to debounce the results so we are not firing off unneeded requests
  // @action
  search = debounce(async () => {
    this.isLoading = true;
    if (!this.searchTerm) {
      this.updateResultSet([]);
    } else {
      const results = await searchRequest(this.searchTerm);
      this.updateResultSet(results.items);
    }
    // We are doing this because we are debouncing the 'change' event and want the loader
    // to show in between the debouncing
    setTimeout(() => this.updateIsLoading(false), 500);
  }, 300);

  @action
  updateResultSet(resultSet) {
    this.resultSet.replace(resultSet);
  }

  @action
  updateIsLoading(isLoading) {
    this.isLoading = isLoading;
  }

  @action
  async getUser(userUrl) {
    this.user = await userInfoRequest(userUrl);
    this.updateResultSet([]);
  }

  @action
  async getFollowers(followersUrl) {
    const { followers, links } = await followersRequest(followersUrl);
    this.followers.replace(followers);
    this.nextPageOfFollowersLink = links ? links.next : null;
  }

  @action
  async getMoreFollowers() {
    this.isLoadingMore = true;
    const { followers, links } = await followersRequest(this.nextPageOfFollowersLink);
    this.followers.replace([...this.followers, ...followers]);
    this.nextPageOfFollowersLink = links.next;
    this.isLoadingMore = false;
  }
}

// We instantiate this here so we only have one instance 
export default new SearchStore();
