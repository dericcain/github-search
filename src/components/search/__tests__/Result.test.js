import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import SearchStore from '../../../stores/SearchStore';
import { Result } from '../';

describe('<Result />', () => {

  // We need to pass the store as a prop and mock our store functions
  let store, mockGetUser, mockGetFollowers, mockResult, resultComponent;

  beforeAll(() => {
    store = SearchStore;
    store.getFollowers = jest.fn();
    store.getUser = jest.fn();
    // Our Result component needs a result prop with these properties 
    // login, avatar_url, url, followers_url
    mockResult = {
      login: 'github.user',
      avatar_url: 'https://google.com',
      avatar_url: 'https://yahoo.com',
    }
    resultComponent = shallow(<Result result={mockResult} searchStore={store} />);
  });

  it('should match the snapshot', () => {
    const resultComponent = renderer.create(<Result result={mockResult} searchStore={store} />).toJSON();
    expect(resultComponent).toMatchSnapshot();
  });

  it('should make a request to get a users info when handleOnClick is called', () => {
    resultComponent.dive().instance().handleOnClick();
    expect(store.getUser).toHaveBeenCalled();
  });

  it('should make a request to get a users followers when handleOnClick is called', () => {
    resultComponent.dive().instance().handleOnClick();
    expect(store.getFollowers).toHaveBeenCalled();
  });
  
});