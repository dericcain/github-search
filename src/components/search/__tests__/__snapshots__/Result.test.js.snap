// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Result /> should match the snapshot 1`] = `
<div
  className="sc-bxivhb isZioG"
  onClick={[Function]}
>
  <img
    className="sc-ifAKCX yqupf"
    src="https://yahoo.com"
  />
  <p
    className="sc-EHOje dYbFBX"
  >
    github.user
  </p>
  <img
    alt="View github.user"
    className="search-result-icon"
    src="search-icon.svg"
  />
</div>
`;
